File: spotlight/src/main/java/com/takusemba/spotlight/shape/Shape.java
package com.takusemba.spotlight.shape;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import com.takusemba.spotlight.target.Target;

/**
 * Shape of a {@link Target}
 * For any shape of target, this Shape class need to be implemented.
 */
public interface Shape {

  /**
   * draw the Shape
   *
   * @param value the animated value from 0 to 1
   */
  void draw(Canvas canvas, PointF point, float value, Paint paint);
}

File: spotlight/src/main/java/com/takusemba/spotlight/shape/Shape.kt
package com.takusemba.spotlight.shape

import android.animation.TimeInterpolator
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import com.takusemba.spotlight.Target

/**
 * Shape of a [Target] that would be drawn by Spotlight View.
 * For any shape of target, this Shape class need to be implemented.
 */
interface Shape {

  /**
   * [duration] to draw Shape.
   */
  val duration: Long

  /**
   * [interpolator] to draw Shape.
   */
  val interpolator: TimeInterpolator

  /**
   * Draws the Shape.
   *
   * @param value the animated value from 0 to 1.
   */
  fun draw(canvas: Canvas, point: PointF, value: Float, paint: Paint)

  /**
   * Checks if point on edge or inside of the Shape.
   *
   * @param anchor center of Shape.
   * @param point point to check against contains.
   * @return true if contains, false - otherwise.
   */
  fun contains(anchor: PointF, point: PointF): Boolean
}
